midterm-2023-11-07/code/5311-kristijan-lazarov/script.ts

```diff
@@ -0,0 +1,322 @@
+interface Movie {
+    id: number
+    title: string
+    year: number
+    director: string
+    genre: string[]
+    plot: string
+    cast: Actor[]
+    oscars: any
+    rating: number
+}
+
+interface Actor {
+    actor: string
+    character: string
+}
+
+
+type MovieSorter = (first: Movie, second: Movie) => number;
+document.addEventListener('DOMContentLoaded', siteCode);
+let movies: Movie[] = [];
+
+const idSorter: MovieSorter = (first, second) => first.id - second.id;
+const nameSorter: MovieSorter = (first, second) => first.title.localeCompare(second.title);
+const directorSorter: MovieSorter = (first, second) => first.director.localeCompare(second.director);
+const yearSorter: MovieSorter = (first, second) => first.year - second.year;
+const oscarSorter: MovieSorter = (first, second) => {
+    const firstOscars = Object.keys(first.oscars).length;
+    const secondOscars = Object.keys(second.oscars).length;
+
+    return secondOscars - firstOscars;
+}
+
+async function siteCode() {
+    const data = await loadData();
+    movies = data;
+
+    fillterGenres(movies);
+
+    displayMovies(movies);
+
+
+    const idSort = document.getElementById('sort-id') as HTMLElement;
+    const nameSort = document.getElementById('sort-name') as HTMLElement;
+    const directorSort = document.getElementById('sort-director') as HTMLElement;
+    const yearSort = document.getElementById('sort-year') as HTMLElement;
+    const oscarSort = document.getElementById('sort-oscar') as HTMLElement;
+
+
+    idSort.addEventListener('click', () => sortByid(idSort));
+    nameSort.addEventListener('click', () => sortByName(nameSort));
+    directorSort.addEventListener('click', () => sortByDirector(directorSort));
+    yearSort.addEventListener('click', () => sortByYear(yearSort));
+    oscarSort.addEventListener('click', () => sortByOscar(oscarSort));
+
+    const applyFilterButton = document.getElementById('apply-filter')!;
+    applyFilterButton.addEventListener('click', applyGenreFilter);
+
+    const nameFilter = document.getElementById('name-filter') as HTMLInputElement;
+    nameFilter.addEventListener('input', applyNameFilter);
+
+    const actorFilter = document.getElementById('actor-filter') as HTMLInputElement;
+    actorFilter.addEventListener('input', applyActorFilter);
+}
+
+
+function updateSortArrow(element: HTMLElement, isAscending: boolean) {
+    const arrows = Array.from(element.querySelectorAll('.sort-arrow')) as HTMLElement[];
+    for (const arrow of arrows) {
+        arrow.style.display = 'none';
+    }
+
+    const arrow = isAscending ? element.querySelector('.sort-arrow.up') as HTMLElement : element.querySelector('.sort-arrow.down') as HTMLElement;
+    if (arrow) {
+        arrow.style.display = 'inline';
+        arrow.innerHTML = isAscending ? '&#129153;' : '&#129155;';
+        arrow.style.color = isAscending ? 'green' : 'red';
+    }
+}
+
+const sortByid = (element: HTMLElement) => {
```

Comment:
All the sort methods have the same structure - we could use a higher order function to actually generate these methods for us